File: typescript/packages/mechanisms/svm/src/mint-cache.ts
import { fetchMint } from "@solana-program/token-2022";
import type { Address } from "@solana/kit";

type Mint = Awaited<ReturnType<typeof fetchMint>>;

export type MintMetadata = {
  decimals: Mint["data"]["decimals"];
  programAddress: Mint["programAddress"];
};

export type MintMetadataCache = Map<string, Promise<MintMetadata>>;

/**
 * Gets stable SPL mint metadata from cache, fetching it once per network and mint.
 *
 * @param rpc - Solana RPC client
 * @param network - Payment requirements network
 * @param asset - SPL mint address
 * @param cache - Per-client mint metadata cache
 * @returns Cached mint metadata
 */
export async function getCachedMintMetadata(
  rpc: Parameters<typeof fetchMint>[0],
  network: string,
  asset: Address,
  cache: MintMetadataCache,
): Promise<MintMetadata> {
  const key = `${network}:${asset}`;
  let metadata = cache.get(key);

  if (!metadata) {
    metadata = fetchMint(rpc, asset).then(mint => ({
      decimals: mint.data.decimals,
      programAddress: mint.programAddress,
    }));
    cache.set(key, metadata);
  }

  try {
    return await metadata;
  } catch (error) {
    if (cache.get(key) === metadata) {
      cache.delete(key);
    }
    throw error;
  }
}

File: python/x402/mechanisms/svm/mint_cache.py
"""Helpers for caching immutable SPL mint metadata."""

from dataclasses import dataclass
from typing import Any

from solders.pubkey import Pubkey

from .constants import TOKEN_2022_PROGRAM_ADDRESS, TOKEN_PROGRAM_ADDRESS


@dataclass(frozen=True)
class MintMetadata:
    """Stable fields needed to build SPL token transfers."""

    decimals: int
    token_program: Pubkey


MintMetadataCache = dict[tuple[str, str], MintMetadata]


def get_cached_mint_metadata(
    client: Any,
    network: str,
    mint: Pubkey,
    cache: MintMetadataCache,
) -> MintMetadata:
    """Get cached mint decimals and token program, fetching once per network/mint."""
    key = (network, str(mint))
    cached = cache.get(key)
    if cached:
        return cached

    mint_info = client.get_account_info(mint)
    if not mint_info.value:
        raise ValueError(f"Token mint not found: {mint}")

    mint_owner = str(mint_info.value.owner)
    if mint_owner == TOKEN_PROGRAM_ADDRESS:
        token_program = Pubkey.from_string(TOKEN_PROGRAM_ADDRESS)
    elif mint_owner == TOKEN_2022_PROGRAM_ADDRESS:
        token_program = Pubkey.from_string(TOKEN_2022_PROGRAM_ADDRESS)
    else:
        raise ValueError(f"Unknown token program: {mint_owner}")

    metadata = MintMetadata(
        decimals=mint_info.value.data[44],
        token_program=token_program,
    )
    cache[key] = metadata
    return metadata
